### PR TITLE
Multi-Context Support

### PR DESCRIPTION
**Pull request by tangledbytes**
**Description**

This PR adds support for multiple contexts in meshery. This is a critical PR, has a **huge** surface area for bugs to pop up and also changes the way sections of meshery server interact with kubernetes cluster.

Details:
1. This PR changes the "unit" of cluster communication from a kubeconfig file to per-user per-cluster kubernetes contexts. Following are a few results of such changes:
    - Any number of kubconfigs can be uploaded to meshery server.
    - Each user gets to have any number of contexts for a meshery instance
    - "Current context" for each user can differ.
    - Long running processes will not be impacted with future context changes of the user.
    - Potential issues (that the author can/could think of):
        - What happens when there is no context?
            - In-cluster deployments - A context is created from the mounted secrets
            - No kubeconfig - no context is created and no handler is created. This does not stops the http flow though, each handler can choose to accept or reject the request. This behaviour was chosen so as to keep meshery working even if no kubernetes cluster is reachable.
       - What if a current-context already exists and a new kubeconfig is uploaded with another "current-context"?
           - Depends on the provider. At the moment, both `None` as well as `Meshery Cloud` will ignore the "current-context" claim of the new context. If the user wants to switch the context, they can do so via the API/UI.
       - What if a "current-context" is deleted? Will there be a new current-context chosen?
           - No
       - Who can see my contexts?
           - Anyone who has access to the meshery instance.
       - Who can modify my contexts?
           - No one. Contexts are immutable.They can be deleted but only by the "owner" of the context.
       - What happens if two users invoke the same API with different contexts?
           - Both of their execution will be isolated. There are no global handlers in meshery server.
       - What happens if two users invoke different Meshery APIs under the same context?
           - Meshery operations remain isolated.
       - What happens if I delete all of my contexts?
           - You are back to the initial stage. Meshery will try to read kubeconfig and find your current-context.
           - Other users will not see those contexts anymore.
           - Inflight operations (yours or those of other users) proceed unaffected and conclude as requested.
       - What happens if the owner of the context has their user account deleted?
           - Context remains. Created_By value goes to nil.
2. Tests done:
    - On local system:
        - With one cluster in kubeconfig:
            - [x] Startup meshery
            - [x] Ping cluster
            - [x] Check details of context in settings
            - [x] Deploy meshery operator
            - [x] Delete current context
            - [x] Deploy service mesh (interact with adapter)
            - [x] Run performance test
        - With two clusters in kubeconfig:
            - [x] Startup meshery
            - [x] Ping each cluster
            - [x] Change context from the UI
            - [x] Check details of context in the settings
            - [x] Delete current context
            - [x] Delete all contexts
            - [x] Deploy meshery operator
            - [x] Deploy service mesh (interact with adapter)
            - [x] Run performance test

**Notes for Reviewers**
Although this PR is functional (to the best of my knowledge) but still it should **not** be merged. Some cleanup items are left to do like:
1. Use meshkit errors

**[Signed commits](https://github.com/meshery/meshery/blob/master/CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [x] Yes, I signed my commits.

### CODE REVIEW

**Comment by utkarsh-pro** [comment]
Reviewers,
@sudo-NithishKarthik, this PR effects UI. Some changes has been done while some still needs to be done. Kindly take a look at it.
@navendu-pottekkat, @piyushsingariya this PR changes the behaviour of some APIs and removes some. This may have implications on mesheryctl. Please look out for them and lemme know.
@bariqhibat this PR changes the way kubehandlers are provided to graphql. Please look out and lemme know if it causes some issues.

**Comment by navendu-pottekkat** [comment]
Thanks for the note @utkarsh-pro Will test out mesheryctl and see if there are any issues.

File: ui/components/MeshConfigComponent.js [hunk 1]
```diff
@@ -711,12 +634,20 @@ handleNATSClick = () => {
           </ListItem>
         </List>
       );
-      if (configuredServer) {
-        chp = <Tooltip title={`Server: ${configuredServer}`}>{chp}</Tooltip>;
-      }
       showConfigured = (
         <div>
-          {chp}
+          {contexts.map(ctx => (
```
Comment (sudo-NithishKarthik) [inline ui/components/MeshConfigComponent.js:639]: we are essentially pinging the kubernetes server right?, all contexts will have the same kubernetes server right. In that case, don't you think we should just have a single chip ? doesn't having multiple chips each with the labels of contexts give us a false idea that the ping is associated with the context ? 
// @leecalcote

Comment (Revolyssup) [reply ui/components/MeshConfigComponent.js:639]: Each kubernetes context is associated with its own kube config and it's own kubernetes server. So essentially, we need multiple chips, one for each Cluster.

Comment (sudo-NithishKarthik) [reply ui/components/MeshConfigComponent.js:639]: Thank you for the clarification @Revolyssup 🙏

**Comment by leecalcote** [comment]
FYI - A conflict cropped up.

**Review by sudo-NithishKarthik — DISMISSED** [review]

**Comment by leecalcote** [comment]
Syncing with master. 

@utkarsh-pro this PR does not include the [Kubernetes context switcher,](https://www.figma.com/file/SMP3zxOjZztdOLtgN4dS2W/Meshery-UI?node-id=3618%3A20324) does it?

Screen Shot 2022-01-04 at 2 08 01 AM

**Comment by leecalcote** [comment]
@utkarsh-pro when multiple contexts are connected, only a single cluster chip shows on the Dashboard. We're expecting to show all connected contexts here, yeah?

https://user-images.githubusercontent.com/7570704/148029630-3401576c-8d9d-4034-8b22-b6aad24ddefd.mov

**Comment by leecalcote** [comment]
@utkarsh-pro I'm getting:

```
INFO[0042] attempting to save kubernetes context to remote provider 
DEBU[0042] saving context to remote provider - constructed URL: https://meshery.layer5.io/user/contexts 
WARN[0042] failed to save the context: Status Code: 201.failed to save kubernetes context 
```

I'll send you the full logs...

**Comment by utkarsh-pro** [comment]
> Syncing with master.
> 
> @utkarsh-pro this PR does not include the [Kubernetes context selector](https://www.figma.com/file/SMP3zxOjZztdOLtgN4dS2W/Meshery-UI?node-id=3618%3A20324), does it?
> 
> Screen Shot 2022-01-04 at 2 08 01 AM
> 
> Without a selector, users will be unable to identify which context a given operation is to be sent to.

@leecalcote, yes this PR doesn't have the mentioned context selector. Right now the only way to switch contexts is to do it from the settings (by selecting the context from the dropdown).

UI certainly needs to be improved, it could be a part of this PR or could be brought in another PR considering that the UX doesn't changes even if we bring in this PR.

**Comment by utkarsh-pro** [comment]
> @utkarsh-pro when multiple contexts are connected, only a single cluster chip shows on the Dashboard. We're expecting to show all connected contexts here, yeah?
> 
>  Screen.Recording.2022-01-04.at.2.17.44.AM.mov

Yes, that's right. We should be showing the same here as well.

**Comment by utkarsh-pro** [comment]
> @utkarsh-pro I'm getting:
> 
> ```
> INFO[0042] attempting to save kubernetes context to remote provider 
> DEBU[0042] saving context to remote provider - constructed URL: https://meshery.layer5.io/user/contexts 
> WARN[0042] failed to save the context: Status Code: 201.failed to save kubernetes context 
> ```
> 
> I'll send you the full logs...

Ah, I think this is just a minor bug. I think the status code that the remote provider sending is 201 (resource created) while the server is expecting 200. So the API is working just fine, just need to change the expected status code.

Will fix this 👍

**Comment by piyushsingariya** [comment]
okay so I ran the PR locally I started two different clusters one with minikube and one with kind, I uploaded the kubeconfig manually too but it doesn't recognize both the clusters at the same time.
image

**Comment by piyushsingariya** [comment]
Things that failed during testing in my system
- Failed to delete current-context
- Failed to delete all context
- Failed to deploy operator in a non-current-context cluster
- If I upload the same kubeconfig twice, multiple entries for contexts get saved

https://user-images.githubusercontent.com/50979887/149617564-757efc3d-4f17-4331-a261-e60f349b666d.mp4

**Review by piyushsingariya — COMMENTED** [review]
Things that failed during testing in my system
- Failed to delete current-context
- Failed to delete all context
- Failed to deploy operator in a non-current-context cluster
- If I upload the same kubeconfig twice, multiple entries for contexts get saved

https://user-images.githubusercontent.com/50979887/149617564-757efc3d-4f17-4331-a261-e60f349b666d.mp4

**Comment by leecalcote** [comment]
@utkarsh-pro thoughts on @piyushsingariya's feedback? 

Heads-up on merge conflicts.

**Comment by utkarsh-pro** [comment]
These are concerning findings @piyushsingariya. I have been away from this PR for a while now but will look into these behaviours and will confirm if these are indeed reproducible bugs and were not left out for certain reasons.

I should be able to get back on this within a few hours.

// @leecalcote

**Comment by utkarsh-pro** [comment]
Also, thank you much @piyushsingariya for testing this out understanding you may had to go out of your way to test this PR out 🙁 .

**Comment by leecalcote** [comment]
@utkarsh-pro here's the design for the Kubernetes context switcher - https://www.figma.com/file/SMP3zxOjZztdOLtgN4dS2W/Meshery-UI?node-id=3984%3A21705

Screen Shot 2022-01-20 at 2 18 13 PM

**Comment by leecalcote** [comment]
@utkarsh-pro any comment here? Something that you can implement?

**Comment by leecalcote** [comment]
A few merge conflicts have popped up. FYI, @utkarsh-pro.

**Comment by utkarsh-pro** [comment]
@leecalcote, yes I can work on the the given UI. Have been lagging quite a lot but will make sure that this PR goes through ✅ .

**Comment by leecalcote** [comment]
@utkarsh-pro sounds great. Can you demo progress at this week's Meshery development meeting?

**Comment by leecalcote** [comment]
@utkarsh-pro, how about this week? Can you demo progress at this week's Meshery development meeting?

**Comment by utkarsh-pro** [comment]
Hey @leecalcote, I have conflicting meeting on Wednesday. I'll try to show the progress on Friday's call though.

**Comment by leecalcote** [comment]
@utkarsh-pro, excellent! // @debo19

**Comment by leecalcote** [comment]
Tomorrow is the big day! :)

**Review by Revolyssup — APPROVED** [review]

**Pull request merged**